ivy/func_wrapper.py

```diff
@@ -1023,43 +1022,52 @@ def _wrap_function(
+                or hasattr(ivy, attr)
+                or attr == "mixed_backend_wrappers"
+            ):
                 continue
             setattr(to_wrap, attr, getattr(original, attr))
         # Copy docstring
         docstring_attr = ["__annotations__", "__doc__"]
         for attr in docstring_attr:
             setattr(to_wrap, attr, getattr(original, attr))
 
-        mixed_fn = (
-            original != to_wrap
-            and hasattr(original, "inputs_to_ivy_arrays")
-            and not original.__name__.startswith("inplace")
-        )
+        mixed_fn = hasattr(original, "mixed_backend_wrappers") and original != to_wrap
+        partial_mixed = mixed_fn and hasattr(to_wrap, "partial_mixed_handler")
+        add_wrappers, skip_wrappers = [], []
+        if mixed_fn:
+            backend_wrappers = getattr(original, "mixed_backend_wrappers")
+            add_wrappers = backend_wrappers.get("to_add")
+            skip_wrappers = backend_wrappers.get("to_skip")
 
         for attr in FN_DECORATORS:
-            if (hasattr(original, attr) and not hasattr(to_wrap, attr)) or (
-                mixed_fn
-                and (
-                    attr
-                    in [
-                        "inputs_to_native_arrays",
-                        "outputs_to_ivy_arrays",
-                        "handle_mixed_function",
-                    ]
-                )
-            ):
-                if mixed_fn:
-                    if attr == "inputs_to_ivy_arrays":
-                        continue
-                    if attr == "handle_mixed_function":
-                        if hasattr(to_wrap, "partial_mixed_handler"):
-                            to_wrap.compos = original
-                            to_wrap = handle_mixed_function(
-                                getattr(to_wrap, "partial_mixed_handler")
-                            )(to_wrap)
-                        continue
-                to_wrap = getattr(ivy, attr)(to_wrap)
+            if hasattr(original, attr) and not hasattr(to_wrap, attr):
+                if attr not in skip_wrappers:
+                    to_wrap = getattr(ivy, attr)(to_wrap)
+
+            elif mixed_fn:
+                if attr == "handle_mixed_function":
+                    if partial_mixed:
+                        to_wrap.compos = original
+                        to_wrap = handle_mixed_function(
+                            getattr(to_wrap, "partial_mixed_handler")
+                        )(to_wrap)
+                    continue
+                if attr in add_wrappers:
+                    to_wrap = getattr(ivy, attr)(to_wrap)
+
+        # we should remove the all the decorators
+        # after handle_mixed_fuction in FN_DECORATORS
+        # from the compos function because these will
+        # be run from the primary implementation.
+        if partial_mixed:
+            array_spec = to_wrap.compos.__dict__["array_spec"]
+            for attr in FN_DECORATORS[-1:14:-1]:
```

Comment:
sorry to be annoying, but we shouldn't be hardcoding with `14` either, it should instead be whatever index `handle_mixed_function` corresponds to 😅

Reply:
Oh, I'm sorry I didn't realize this. Thanks a lot for pointing this out. Have the updated the function now :smile: